External processing filter protos

gbrail opened the pull request:
This introduces a new filter called the "external processing filter." It is intended to allow an external service to be able to operate as if it were part of the filter chain using a gRPC stream. It is intended to support a variety of use cases in which processing of HTTP requests and responses by an external service is desired.

A document that describes the filter can be found here:

https://docs.google.com/document/d/1IZqm5IUnG9gc2VqwGaN5C2TZAD9_QbsY9Vvy5vr9Zmw/edit#heading=h.3zlthggr9vvv

gbrail commented:
This replaces

https://github.com/envoyproxy/envoy/pull/13304

which was developing a very long comment stream.

mattklein123 commented:
@gbrail how do you want to handle my last round of comments from the other PR? Do you want to take a stab at them and then I can look again?

/wait-any

gbrail commented:
I don't know if I can open that PR again (I'll try again today) but I addressed them in the version of the protos that is shown here.

If you look in particular at HeaderResponse and BodyResponse, they now include additional fields so that the remote server can completely replace the message body and trailers on either the request or response path, and they come with comments that purport to explain what will happen. This allows adding a message body to one that didn't have one, and removing the body as well.

What I don't think we have is a way to replicate the exact behavior of "ContinueAndDontEndStream," in which Envoy keeps processing the filters but expects a new body later. We also don't have a way for the remote service to choose between "StopAllIterationAndWatermark" and "StopAllIterationAndBuffer" while waiting for the service to respond. If we need to add that level of control then we probably need another iteration of the protocol.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,300 @@
+// have not already been sent downstream, then this will cause the filter to
+// immediately return the data specified here to the downstream
+// system, stop further filter processing, and ignore further messages coming
+// from the remote server. However, if the response headers have already been
+// delivered downstream, then this message will be ignored.
```
I think you need way to do a reset though? Do you want to have an explicit reset message, or do you want to use this also as the reset message?

gbrail replied:
Let's make this consistent with "sendLocalReply" on the HTTP filter interface. I'll add the details and change to a "HeaderMutation" structure.

However, FWIW, sendLocalReply doesn't let you set trailers, only the grpc-status. I think I will make that change as well.

gbrail replied:
And by "make this consistent," I mean that sendLocalReply will respond with a new HTTP response if it can and reset the stream if it's too late for that -- I'll adopt that behavior and likely the comment text as well.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,300 @@
+// system, stop further filter processing, and ignore further messages coming
+// from the remote server. However, if the response headers have already been
+// delivered downstream, then this message will be ignored.
+message ImmediateResponse {
```
I think you probably want to also add a local reply reason details for logs.

gbrail replied:
Ack.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,300 @@
+  // received from downstream will be ignored, and any additional messages
+  // from the remote server on either the request or response path will
+  // also be ignored for the duration of this request or response.
+  bool end_of_stream = 4;
```
I think you probably want this to be a WKT so you have a tri-state. The reason for that is that I think you also want to be able to set end_stream to false, and make a headers only request/response into a request/response with data, by sending data/trailers later. I commented on this in the other PR. I don't think we have to support this advanced functionality on day 1, but it would be nice to not have to deprecate things to do it. The way I think it would end up working is end_of_stream would be explicitly set to false (as opposed to empty being no change), and then the server could stream a sequence of body messages and an optional trailer message to finish the request/response. Let me know if this makes sense.

gbrail replied:
I don't mind adding the tri-state, or perhaps another flag to make the protocol more general -- let me think about it.

In the meantime, here's how this particular use case might work:

1. onRequestHeaders is called and the "end_stream" flag is set.
2. filter sends the request_headers message on the gRPC stream.
3. Filter returns with StopAllIterationAndWatermark. (Because we don't know what the remote server wants yet we can't return a different code IMO.)
4. Filter receives a message on the stream with a status of "end_stream_later" or whatever we decide to call it
5. Filter continues to wait for additional responses with an "end_stream_finally" message. In the meantime, if it receives multiple message body messages, it buffers them.
6. Filter gets the "end_stream_finally" message and then calls continueDecoding() to resume the filter chain

Note that this is different from the ContinueAndDontEndStream status that's in the HTTP filter interface, because with that interface, we continue processing the filter chain and wait for the new data, whereas here we can't continue processing in that way because we already returned a different status and the filter interface gives us no way to control what "continueDecoding()" means.

I might argue that unless the filter processing options change, we could instead roll up steps 3-5 there into a single message -- let the remote server do all that work and respond only when it's ready for Envoy to continue processing.

mattklein123 replied:
I think you have the right idea, but I think it's possible to also stream data from the ext proc server without buffering, and I could see this being desired in advanced use cases. See the "inject" family of functions in filter.h.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,300 @@
+  HeaderMutation header_mutation = 1;
+}
+
+// TODO this is now the same as HeaderResponse. However, do we think
+// that will always be the case? Keeping it as a separate message
+// for now.
```
You might consider a common shared message and a wrapper to allow for extensibility.

gbrail replied:
I'll take a look later today or tomorrow.

mattklein123 reviewed (changes requested):
Overall this LGTM modulo a few more comments. @htuch @alyssawilk @PiotrSikora do you want to take another pass? Thanks for all the iteration on this!

/wait

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -182,99 +187,111 @@ message HttpTrailers {
+    // headers. If body_mutation is set and contains a body, then add that
+    // body to the request or response, even if one does not already exist --
+    // otherwise, clear the body. Any additional body and trailers
+    // received from downstream will be ignored, and any additional messages
```
```suggestion
    // received from downstream or upstream will be ignored, and any additional messages
```

gbrail replied:
Ack.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -182,99 +187,111 @@ message HttpTrailers {
+    // body to the request or response, even if one does not already exist --
+    // otherwise, clear the body. Any additional body and trailers
+    // received from downstream will be ignored, and any additional messages
+    // from the remote server on either the request or response path will
+    // also be ignored for the duration of this request or response.
```
I think this text is wrong? Isn't the point of this that you want the remove server to be able to send more data, but have it be streamed?

gbrail replied:
I was trying to distinguish this use case from the "pause and do not end stream" status. But I'd rather just take that status out and simplify things a bit.

mattklein123 replied:
I'm not sure I follow. With how this is written I'm not sure what would ever continue/finish the stream?

gbrail replied:
I can see how that reads confusingly and I'm going to take it out. I think it's clear enough without it.

mattklein123 commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -182,99 +187,111 @@ message HttpTrailers {
+    // have one already.
+    CONTINUE_AND_DO_NOT_END_STREAM = 1;
+
+    // Replace the request or response as with CONTINUE_AND_DO_NOT_END_STREAM,
+    // but do not continue filter processing yet -- the remote server can
+    // send additional ProcessingResponse messages with this value
+    // set that further modify the message or add new data, and
+    // must eventually reply with one of the "CONTINUE" statuses
+    // in order for filter processing to resume.
```
I would make it clear that this results in buffering? Stepping back, is there any real reason that we would ever stream chunks from the ext proc server and buffer it in the filter? Can this all be simplified into either CONTINUE or CONTINUE_AND_DO_NOT_END_STREAM?

gbrail replied:
I think in the long conversation that we had last week, we talked about a case -- adding a body to a message with no body already -- in which the filter chain allows a filter to insert data to the filter chain, and there's that specific return value from the "on*Headers" callback that's used for it. So this return value supports that.

However, given that the filter can't actually return that special error code, I'd certainly rather simplify this thing by only having two states -- continue and continue_and_do_not_end_stream, because the ext_proc server can always buffer. Then of course we can add another one in the future.

If you're OK with that I'll make it.

mattklein123 replied:
Yes, I think this makes the most sense for now.

mattklein123 reviewed (changes requested):
LGTM with 1 remaining comment. Thank you!

/wait

gbrail reviewed (commented):
Great! Let's just resolve what the "ResponseStatus" enum means, I'll update it, and re-submit if you agree that two states are OK for now.

mattklein123 reviewed (approved):
LGTM, over to other @envoyproxy/api-shepherds. Great work! Once everyone approves of the API let's circle back and discuss what PR sequence we are actually going to use to implement this?

htuch commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,145 @@
+  // Envoy will lookup the runtime key to get the percentage of requests to filter.
+  //
+  // If this field is not specified, the filter will be enabled for all requests.
+  config.core.v3.RuntimeFractionalPercent filter_enabled = 8;
```
Should something like this really be promoted to HTTP filter wrapper message @envoyproxy/api-shepherds? It seems like this is something that should be implemented once in HCM, rather than each filter.

gbrail replied:
I'd of course be happy if this weren't something I have to copy from another filter. Let me know if you think I should leave this field out for now and plan to add it back later.

htuch replied:
I'd suggest eliding it for now and filing an issue for a generic HCM fractional percent filter enablement. We can always add it back here if needed while that issue is pending.

gbrail replied:
Ack -- I will take it out of this PR.

htuch commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,145 @@
+  ProcessingMode processing_mode = 1;
+
+  // Set a different asynchronous processing option than the default.
+  bool async_mode = 2;
```
Would it make sense to factor these out with the same settings in the above message?

gbrail replied:
Previously we had an option for the ext_proc server to change the ProcessingMode either permanently or for a certain duration so that the filter could switch between async and sync mode while running. After some discussion we agreed that this extra complexity would be confusing and obviously a race condition. So we changed it so that now ProcessingMode, which specifies which callbacks result in a message to the remote server, is overridable on a per-message basis, but the only way to change async mode is via configuration.

So since ProcessingMode can be overridden for each message, but async mode can only be changed via a configuration change, I think that they need to be separate.

htuch commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,307 @@
+    // received from downstream or upstream will be ignored.
+    // This can be used to add a body to a request or response that does not
+    // have one already.
+    CONTINUE_AND_DO_NOT_END_STREAM = 1;
```
I'm having a hard time parsing this name in contrast to `CONTINUE` above. Does the above `CONTINUE` end the stream and this one doesn't? This seems a little odd.

gbrail replied:
The intention of this filter was to support the same kind of workflow as the ContinueAndDontEndStream status in FilterHeadersStatus, which is used to add a body and trailers to an HTTP request that does not have them. So I copied the name. But if you have an idea for a better name then that'd be great!

htuch replied:
Envoys internal naming isn't great as it's not a stable public API, I think in this case we shoul

How about `REPLACE`?

What does Wasm do here @PiotrSikora ?

gbrail replied:
I'm changing it to CONTINUE_AND_REPLACE, which makes it clear that we're still continuing processing.

PiotrSikora replied:
We don't have this case covered in Wasm right now, but to be honest, I'm not a big fan of mixing control of the processing flow (stop/continue) with mutating `end_stream` value.

mattklein123 replied:
The original intention of this back and forth was to allow a scenario in which we:
1) Tell the filter to continue but not end the stream.
2) Later send more messages to eventually complete the stream.

At it has been updated and now written, I agree that CONTINUE_AND_REPLACE is what makes sense, but we should make it very clear this is basically END_STREAM_AND_CONTINUE or CONTINUE_AND_END_STREAM (would that be better?)

In the future we can add CONTINUE_AND_DONT_END_STREAM

PiotrSikora replied:
The issue is that we have 3 options here:

1) `Continue` (and don't modify `end_stream`),
2) `ContinueAndEndStream` (it was removed last month in #13216, but the justification seems wrong),
3) `ContinueAndDontEndStream`.

and we really should have the same values available in `FilterDataStatus`, since we might want to truncate body and/or replace it with something else, which ends up with a weird interface where request/response body is mutated one way, and the `end_stream` flag for the same request/response is mutated in a different way.

I think it would be useful if we could decouple those 2 and limit use of return values to control processing flow (i.e. continue/stop), and mutate request/response body or network bytestream via APIs.

mattklein123 replied:
I agree, and these are the cases I was pushing for when we were iterating on this PR. I agree the C++ API is suboptimal, but it is what it is and has evolved over time.

I want to make sure we can support all of these cases in both WASM and with ext_proc so IMO we should focus more on making sure we can support these cases eventually, even if we don't do it now, and worry less about the C++ implementation.

htuch commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,307 @@
+
+// This message must be sent in response to an HttpBody message.
+message BodyResponse {
+  CommonResponse response = 1;
```
Why do `HeadersReponse` and `BodyResponse` devolve to the same `CommonResponse`? It seems that they would have different structure, e.g. you can't modify headers in body response after headers have been sent.

gbrail replied:
Well sometimes you can and sometimes you can't. You can't modify headers in the body response after they have been sent but I think that you can if you buffered the body (at least that's something I want to support, so that an ext_proc server can modify everything together.) You could also in theory decide in a body callback to ignore the rest of the body from the upstream and downstream and ignore the headers and replace the rest of the HTTP message with your own thing -- that's what the CONTINUE_AND_DONT_END_STREAM status is supposed to do.

htuch replied:
It seems that we're linearizing Envoy's internal filter callbacks here, since HTTP filters essentially treat headers and body as independent event streams. Is this by design?

gbrail replied:
The protocol tries to keep the request and response streams separate -- even though there are separate HttpHeaders and HttpBody messages, they're broken into separate fields like request_headers and response_headers for that reason. So the design is not to linearize that.

However, I am assuming that the request or response headers will come before the body, and I'm planning to use "StopAllIterationAndWatermark" so that the ext_proc server has a chance to respond to the request and response headers callbacks before being asked to present the message. If that's not valid then I'm not sure what to do, because again I'm sure that we'll want the ability to modify the headers and body together sometimes.

htuch replied:
Make sense. Looking at the request flow, I'm trying to understand how this works though. Let's say you have sent to the server two messages, one with headers, and the next with body. How many response messages are expected? One for both the headers and body request message? One just for the body request message? Would an empty headers response message be sent when you want to do processing in the body?

Do we have a strict request-response protocol that doesn't allow pipelining? Would it be more efficient for Envoy to be able to fire the headers and body requests without feedback?

gbrail replied:
That's what's documented right now -- Envoy sends a "request_headers" message, pauses the filter chain, and waits for a "request_headers" response. That seemed like the most straightforward thing for an ext_proc server writer to support -- then the server can respond with a new processing mode if it wants the body messages, or not. Otherwise it's already too late to respond I think.

If we want to support other modes, we can add more options to the ProcessingMode message or fine-tune our interpretation there.

htuch replied:
It's probably fine to start with this, but I'd be curious what it would take to reduce the round trips in a later API. When folks start optimizing latency of ext_proc based requests, this is probably going to be low hanging fruit.

htuch commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,307 @@
+  // along with the CONTINUE_AND_DO_NOT_END_STREAM status.
+  config.core.v3.HeaderMap trailers = 4;
+
+  // Clear the route cache for the current request.
```
From a wire perspective, I think this is leaking Envoy implementation details out. Should this be something more imperative, e.g. `recalculate_route`? CC @markdroth in case gRPC ever implements something like this protocol.

gbrail replied:
That's the same name that we use in the ext_proc filter, but another name is fine with me.

htuch reviewed (commented):
This is looking great! I have a few questions/comments. Please forgive if they cover well trodden ground, consider my pass over the PR to be with a fresh set of eyes and from the perspective if a new user.

htuch commented on api/envoy/service/ext_proc/v3alpha/external_processor.proto:
```diff
@@ -0,0 +1,307 @@
+  // The values of properties selected by the "request_properties"
+  // list in the configuration. Each entry in the map is the
+  // value of the specified property.
+  map<string, google.protobuf.Struct> properties = 2;
```
FWIW, when https://github.com/envoyproxy/envoy/pull/13971 merges, it would be a good idea to add some changes explaining the ext_proc encoding for the properties (there is already verbiage there for CEL and Wasm).

gbrail replied:
Makes sense to add the cross reference when it's available.

htuch replied:
It is now :)

htuch commented:
@gbrail what's the status of this PR? Are you looking for further review input or are you working on implementation?

gbrail commented:
I believe I have addressed the comments above. I think that this PR is ready to go. Right now, it contains a do-nothing implementation. I plan to implement it in pieces, starting next with just the ability to send a gRPC to the remote service that manipulates the HTTP request and response headers, and then adding more bells and whistles as we go.

Can you all please let me know the next steps to approval? Thanks!

htuch commented on api/envoy/extensions/filters/http/ext_proc/v3alpha/ext_proc.proto:
```diff
@@ -0,0 +1,150 @@
+  // or could not be opened, processing continues without error.
+  bool failure_mode_allow = 2;
+
+  // [#not-implemented-hide:]
```
I would recommend just putting this annotation at message level everywhere for now.

gbrail replied:
I added it to all the fields in config, and hope to be able to unlock a few in a small number of PRs.

htuch replied:
OK, if you prefer doing it that way.

htuch commented on source/extensions/filters/http/ext_proc/BUILD:
```diff
@@ -0,0 +1,34 @@
+    name = "config",
+    srcs = ["config.cc"],
+    hdrs = ["config.h"],
+    security_posture = "unknown",
```
Add `status = "alpha"` as well for now.

gbrail replied:
Ack.

htuch reviewed (commented):
I'm good to ship this given we have agreement on the macro-discussion with the understanding we might still want to iterate on the API until we're ready to consider it stable. LGTM modulo minor comments.

@envoyproxy/api-shepherds please chip in if you are not in agreement with this plan.

mattklein123 commented:
Yeah I'm good to ship and iterate as long as everything is marked alpha and we can change things without deprecation. 

@gbrail please make a note to actually make this not-alpha before this goes into prod somewhere. We have had that issue multiple times and I would like to not repeat it here. :)

gbrail commented:
Great -- I think that this is ready to go as well. Thanks!

gbrail commented:
I will also be careful to mark this not-alpha once it is reasonably feature complete and you all are OK with it. Let's not make it "alpha" forever or else I share your concern about it going live too soon.

htuch reviewed (approved):
LGTM, thanks!
